addons/mail/models/mail_activity.py

```diff
@@ -601,6 +601,16 @@ class MailActivityMixin(models.AbstractModel):
         related='activity_ids.summary', readonly=False,
         search='_search_activity_summary',
         groups="base.group_user",)
+    exception_activity_ids = fields.One2many(
+        'mail.activity', 'res_id', 'Exception Activities',
+        auto_join=True,
+        groups="base.group_user",
+        domain=lambda self: [('res_model', '=', self._name), ('activity_type_id.decoration_type', 'in', ['warning', 'danger'])])
+    exception_activity_type = fields.Char(string=' ', compute='_compute_exception_activity_type')
+
+    def _compute_exception_activity_type(self):
```

Comment:
After a quick look, I would remove the exception_activity_ids field, base the exception_activity_type on activity_ids.
It could be a stored computed field BUT invalidation could be expensive since the dependency is very specific. 

I think that the most effective solution would be to do that manually, just add a exception_activity_type field on mixin and maintains his value on activity write/create 
pseudo code: 
write/create/unlink for activity:
if self.env[model].browse('res_id') .is_mail_activity and  activity_type in values or active in values:
      self.env[model].browse('res_id') ._compute_exception_activity_type

_compute_exception_activity_type should be based on activity_type with filters instead. 

Just a quick draft, feel free to adapt names, logic.